lib/src/query/filter.dart

```diff
@@ -1,35 +1,93 @@
 import 'dart:collection';
 
-class Filter with MapMixin<String, String> {
+class Filter with MapMixin<String, Object> {
   /// Example:
   /// ```dart
-  /// Filter({'post': '1,2', 'author': '12'}).addTo(url);
+  /// Filter({'post': '1,2', 'author': {'id': '12', 'role': 'admin'}}).addTo(url);
   /// ```
   /// encodes into
   /// ```
-  /// ?filter[post]=1,2&filter[author]=12
+  /// ?filter[post]=1,2&filter[author][id]=12&filter[author][role]=admin
   /// ```
-  Filter([Map<String, String> parameters = const {}]) {
+  Filter([Map<String, Object> parameters = const {}]) {
     addAll(parameters);
   }
 
-  static Filter fromUri(Uri uri) => Filter(uri.queryParametersAll
-      .map((k, v) => MapEntry(_regex.firstMatch(k)?.group(1) ?? '', v.last))
-    ..removeWhere((k, v) => k.isEmpty));
+  static Filter fromUri(Uri uri) {
+    final filters = <String, Object>{};
+    uri.queryParametersAll.forEach((key, value) {
+      if (_validationRegex.hasMatch(key)) {
+        final matches = _extractionRegex.allMatches(key).toList();
+        _convertToMapAndMerge(matches, filters, value.last);
+      }
+    });
+    return Filter(filters);
+  }
+
+  static void _convertToMapAndMerge(
+    List<RegExpMatch> matches,
+    Map<String, Object> destination,
+    String value,
+  ) {
+    final key = matches[0].group(1) ?? '';
+    if (key.isEmpty) {
+      return;
+    }
+    if (matches.length == 1) {
+      destination[key] = value;
+      return;
+    }
+    if (!destination.containsKey(key) ||
+        destination[key] is! Map<String, Object>) {
+      destination[key] = <String, Object>{};
+    }
+    _convertToMapAndMerge(
+      matches.sublist(1),
+      destination[key] as Map<String, Object>,
+      value,
+    );
+  }
 
-  static final _regex = RegExp(r'^filter\[(.+)\]$');
+  static final _validationRegex = RegExp(r'^filter(?:\[[^\[\]]+\])+$');
```

Comment:
Do we need it non-capturing (`?:`)?

Reply:
I don't think it really matters if it's capturing or not, because this particular regex is here just to make sure the input string is of fomat `filter[abc][abc]...`, it doesn't actually extract anything. The second regex is for extracting data so it's using capturing groups.
I couldn't figure out a regex which would validate the filter part and extract a variable amount of brackets in one go.